### PR TITLE
Implement yaml extensions overwriting the default pod/container spec

### PR DESCRIPTION
**Pull request by nikola-jokic**
Implements [ADR](https://github.com/actions/runner-container-hooks/pull/96)

### CODE REVIEW

**Comment by fhammerl** [comment]
When running `docker` hooks with {...} `options`, or `k8s` hooks with a string `options`, `options` is ignored as it cannot be parsed.
Write to logs when options is ignored.

File: packages/k8s/src/k8s/utils.ts [hunk 1]
```diff
@@ -158,6 +158,91 @@ export function generateContainerName(image: string): string {
...
+  base: k8s.V1PodSpec,
+  from: k8s.V1PodSpec
+): k8s.V1PodSpec {
+  const newPodSpec = JSON.parse(JSON.stringify(base)) as k8s.V1PodSpec
+
+  for (const [key, value] of Object.entries(from)) {
+    if (key === 'container' || key === 'containers') {
+      continue
+    } else if (key === 'volumes' && value) {
+      const volumes = value as k8s.V1Volume[]
```
Comment (fhammerl) [inline packages/k8s/src/k8s/utils.ts:228]: NIT: Can we extract this into a small function `appendVolumes`?

Comment (nikola-jokic) [reply packages/k8s/src/k8s/utils.ts:209]: Created a generic function because we need it for envs, volumeMounts and ports

**Comment by cjreyn** [comment]
Hi. Is there any ETA for this PR to be merged? We're eagerly awaiting it as I understand it will allow us to request K8s resources such as GPUs for the build jobs.

**Review by fhammerl — CHANGES_REQUESTED** [review]
Let's add an example .yaml that shows some possible overrides

File: packages/k8s/tests/run-container-step-test.ts [hunk 2]
```diff
@@ -23,6 +27,47 @@ describe('Run container step', () => {
     expect(exitCode).toBe(0)
   })
 
+  it('should run pod with extensions applied', async () => {
```
Comment (fhammerl) [inline packages/k8s/tests/run-container-step-test.ts:30]: leave a `//` explaining that we only assert successful run here not individual overrides

File: packages/k8s/tests/k8s-utils-test.ts [hunk 3]
```diff
@@ -328,4 +333,183 @@ describe('k8s utils', () => {
...
+      ports: [
+        {
+          containerPort: 8080,
+          hostPort: 8080,
+          protocol: 'TCP'
+        }
+      ]
+    } as k8s.V1Container
+
+    const from = {
```
Comment (fhammerl) [inline packages/k8s/tests/k8s-utils-test.ts:414]: NIT: maybe `from` -> `extension`

File: packages/k8s/src/k8s/utils.ts [hunk 4]
```diff
@@ -158,6 +162,97 @@ export function generateContainerName(image: string): string {
...
+// to append more values
+//
+// Rest of the fields are just applied
+// For example, container.createOptions.container.image is going to overwrite container.image field
+export function mergeContainerWithOptions(
+  base: k8s.V1Container,
+  from: k8s.V1Container
+): void {
+  for (const [key, value] of Object.entries(from)) {
+    if (key === 'name') {
```
Comment (fhammerl) [inline packages/k8s/src/k8s/utils.ts:178]: NIT: should we ` || key === 'image` here?

File: packages/k8s/src/k8s/utils.ts [hunk 5]
```diff
@@ -158,6 +162,97 @@ export function generateContainerName(image: string): string {
...
+  }
+}
+
+export function mergePodSpecWithOptions(
+  base: k8s.V1PodSpec,
+  from: k8s.V1PodSpec
+): void {
+  for (const [key, value] of Object.entries(from)) {
+    if (key === 'containers') {
+      base.containers.push(
```
Comment (fhammerl) [inline packages/k8s/src/k8s/utils.ts:202]: NIT: extract function `mergeContainers` like a few lines below with `mergeLists` ?

File: examples/extension.yaml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+metadata:
```
Comment (fhammerl) [inline examples/extension.yaml:1]: NIT: comments indicating section by section if it's an 'override' or a 'merge'?

File: packages/k8s/src/k8s/index.ts [hunk 7]
```diff
@@ -137,6 +152,16 @@ export async function createJob(
     }
   ]
 
+  if (extension) {
+    if (extension.metadata) {
+      mergeObjectMeta(job, extension.metadata)
```
Comment (fhammerl) [inline packages/k8s/src/k8s/index.ts:157]: leave `//` with explaining how a container step is wrapped into a k8s job and we want to apply the metadata extensions to both

**Review by fhammerl — APPROVED** [review]
LGTM!

**Comment by cjreyn** [comment]
A huge thanks for your work on this! We plan on using it for a build farm at Diamond Light Source (the UK's national synchrotron). So you are directly helping us with valuable scientific research into batteries, drugs for cancer, and materials science for renewables.

**Comment by nikola-jokic** [comment]
Hey @cjreyn,

Sorry that we did not respond earlier! Thank you for using the container hook and having interest in it! We are glad if it helps you with your work!

**Comment by nielstenboom** [comment]
Happy to see this merged! Seems you went with the template file after all 😉 

I'll close my PR here: https://github.com/actions/runner-container-hooks/pull/50

Are there also plans for integrating this change in the actions-runner-controller downstream? We internally would love to move away from our current complex fork 😄

**Pull request merged**